### DISCUSSION TITLE
microsoft::ui::colors and windows::ui::color

### DISCUSSION THREAD

**Discussion by mgradwohl** [microsoft/WindowsAppSDK#3199] [Q&A]
I came across the colors confusion today. See https://mariusbancila.ro/blog/2022/04/08/unwrapping-winui3-for-cpp/ for someone else's experience.

Why isn't there a microsoft::ui::color living in the same namespace as fromargb and colors?

**Comment by DarranRowe** [comment]
The two curious things about this are:
1) Microsoft.UI.Colors and Microsoft.UI.ColorHelper return Windows.UI.Color objects. This means that there is no difference between the helpers in the Microsoft.UI namespace and the Windows.UI namespace.
2) The WinUI 3 objects that take a colour as a parameter take objects of type Windows.UI.Color.

If you do not believe this, you can find all of this information in the metadata (.winmd files). These are viewable with ildasm and this comes with the .NET stuff in the Windows SDK.

What's more, [the documentation](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api) does state that Windows.UI.Colors and Windows.UI.ColorHelper are usable from desktop applications. Or rather doesn't state that it is unusable with the implication that all are usable unless stated otherwise.

So, for the direct answer, there doesn't need to be a Microsoft.UI.Color and the WinUI team apparently decided to use the existing Windows.UI.Color. I appreciate that this can cause confusion though, especially with the prevailing "stuff in the UWP namespaces can only be used in UWP applications" that persists even now. This has never been true.

Initially, Microsoft did try to use the DualApiPartition attribute to mark things which were usable from desktop applications, but as Windows evolved and the concept of desktop application changed, this became more difficult. There are also things in the UWP namespaces that can only be used from desktop applications too. As an example, the [DesktopWindowTarget](https://learn.microsoft.com/en-us/uwp/api/windows.ui.composition.desktop.desktopwindowtarget?view=winrt-22621) runtime class can be used to enable composition on a Windows API HWND and only a Windows API HWND.

**Comment by MikeHillberg** [comment]
It is confusing. This was the thinking behind it ... 

There's two SDKs: the Platform SDK (ships with the OS) and the App SDK (ships on nuget.org and Microsoft Store). We took many APIs in the Platform SDK and put them in the WinAppSDK, including much of the Windows.UI namespace. We didn't move [Color](https://docs.microsoft.com/uwp/api/Windows.UI.Color) because it's used by types that are only in the Platform SDK, such as media & clipboard APIs (e.g. [DataPackagePropertySet](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.DataTransfer.DataPackagePropertySet)).

We also didn't see any benefit to moving Color, because we don't intend to ever change it anyway. We _did_ foresee updating [Colors](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Colors) and [ColorHelper](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.ColorHelper) though over time, which is why that did get put into the WinAppSDK.

The key unfortunate confusion this creates is that now there's motivation to have using statements in your code for both Windows.UI and Microsoft.UI, which creates a conflict with a few things such as [WindowId](https://docs.microsoft.com/uwp/api/Windows.UI.WindowId) in the Platform SDK and [WindowId](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.WindowId) in WinAppSDK.

**Reply by jaigak** [reply]
Microsoft.UI.ColorHelper doesn't even seem to have ToDisplayName, unlike Windows.UI.Color, so the only remaining method is FromArgb but that's useless and shouldn't - just use aggregate initialisation, much more efficient than making a WinRT static method call.
